Port SetAcceptance to the TCK

petraselmer opened the pull request:
This must only be merged when https://github.com/neo4j/neo4j/pull/7366 is merged

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | node | 42 |
+    When executing query:
+      """
+      SET (CASE WHEN true THEN {node} END).name = 'neo4j'
```
CASE expressions are not openCypher scope.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+Feature: SetAcceptance
+
+  Scenario: Setting a node property to null removes the existing property
+    Given any graph
```
I don't think we can say `any` when we have a `MATCH` in the tested query.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | n                    |
+      | (:A {property2: 46}) |
+    And the side effects should be:
+      | +properties | 1 |
```
I thought the side effects would've been one negative as well?

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | r                      |
+      | [:REL {property2: 24}] |
+    And the side effects should be:
+      | +properties | 1 |
```
The side effect should be just `-properties | 1`, I think.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +properties | 1 |
+
+  Scenario: Setting a relationship property to null removes the existing property
+    Given any graph
```
Can't have `any graph` if we're matching.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +properties | 1 |
+
+  Scenario: Set a property to an expression
+    Given any graph
```
Can't have `any graph` if we're matching; same for all below scenarios.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+    When executing query:
+      """
+      MATCH (n:A)
+      SET (CASE WHEN true THEN n END).name = 'neo4j'
```
No CASE expressions in oC.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+    When executing query:
+      """
+      MATCH (n:A)
+      SET (n).name = 'neo4j'
```
Are the parentheses intentional here?

petraselmer replied:
@Mats-SX yes, the original test has this query: https://github.com/neo4j/neo4j/blob/3.1/community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/SetAcceptanceTest.scala#L80

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+    When executing query:
+      """
+      MATCH ()-[r:REL]->()
+      SET (CASE WHEN true THEN r END).name = 'neo4j'
```
No CASE expressions in oC.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | n              |
+      | (:A {age: 35}) |
+    And the side effects should be:
+      | +properties | 1 |
```
Should be a negative effect.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+    When executing query:
+      """
+      MATCH (n:A)
+      SET n:FOO
```
The label should be `Foo`.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      """
+      MATCH (n:A)
+      SET n.x = [1, 2, 3]
+      RETURN extract(i IN n.x | i/2.0) AS x
```
The `/` operator should have padding space around it.

Mats-SX replied:
Also, no `extract()`. Use a list comprehension, like so:

`RETURN [i IN n.x | i / 2.0]`

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +properties | 1 |
+
+  Scenario: Set a property by selecting the node using a simple expression
+    Given any graph
```
Can't have `any graph` if we're matching.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +properties | 1 |
+
+  Scenario: Set a property by selecting the node using an expression
+    Given any graph
```
Can't have `any graph` if we're matching.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +properties | 1 |
+
+  Scenario: Set a property by selecting the relationship using a simple expression
+    Given any graph
```
Can't have `any graph` if we're matching.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+    When executing query:
+      """
+      MATCH ()-[r:REL]->()
+      SET (r).name = 'neo4j'
```
Are the parentheses intentional here?

petraselmer replied:
@Mats-SX yes, the original query is here: https://github.com/neo4j/neo4j/blob/3.1/community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/SetAcceptanceTest.scala#L104

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +properties | 1 |
+
+  Scenario: Set a property by selecting the relationship using an expression
+    Given any graph
```
Can't have `any graph` if we're matching.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +properties | 1 |
+
+  Scenario: Setting a property to null removes the property
+    Given any graph
```
Can't have `any graph` if we're matching.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +properties | 1 |
+
+  Scenario: Add a label to a node
+    Given any graph
```
Can't have `any graph` if we're matching.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +labels | 1 |
+
+  Scenario: Adding a list property
+    Given any graph
```
Can't have `any graph` if we're matching.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +properties | 1 |
+
+  Scenario: Retain old values when using +=
+    Given any graph
```
Can't have `any graph` if we're matching.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +properties | 2 |
+
+  Scenario: Overwrite values when using +=
+    Given any graph
```
Can't have `any graph` if we're matching.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | n               |
+      | (:X {bar: 'B'}) |
+    And the side effects should be:
+      | +properties | 1 |
```
Should be a negative effect.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | n                         |
+      | (:X {foo: 'B', baz: 'C'}) |
+    And the side effects should be:
+      | +properties | 3 |
```
Two + and one -

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      SET (CASE WHEN true THEN {node} END).name = 'neo4j'
+      RETURN count(*)
+      """
+    Then a ArgumentError should be raised at runtime: RequiresNodeOrRelationship
```
The problem is that the property access operator (`.`) does not work on the boolean value, right? That sounds to me like `IllegalArgumentValue`. A map would also be a valid value, and having `RequiresXorYorZ` type error details quickly gets out of hand I believe.

Mats-SX commented on tck/features/SetAcceptance.feature:
```diff
@@ -0,0 +1,332 @@
+      | +properties | 1 |
+
+  Scenario: Explicit null values in a map remove old values
+    Given any graph
```
Can't have `any graph` if we're matching.

petraselmer commented:
@Mats-SX thanks for reviewing. I have removed all tests containing `CASE`; these ought to be simulated and added later on in one iteration. Also, have hopefully addressed all of your other comments.
